### ISSUE TITLE
Keep third-person camera from going through walls

### ISSUE THREAD

**Issue by TranquilMarmot** [ramokz/phantom-camera#410]
### Issue description

Right now with the third-person camera, even with a spring arm collision shape set, the camera will clip through walls:

From the `3d_follow_third_person_example_scene.tscn` demo:
<img width="348" alt="Screenshot 2024-10-25 at 12 02 26 AM" src="https://github.com/user-attachments/assets/754ac971-4c75-4514-b63f-add29d2ad09c">

https://github.com/user-attachments/assets/44754ec5-baae-478d-9df1-0a69197e006a

Is there a way to avoid this? I tried setting various collision shapes for the spring arm but they don't really have much effect.

At this point I'm considering a ray cast from the camera out and the adjusting the follow offset if it hits anything, but that feels like a hack that the spring arm should already handle.

### Steps to reproduce

Repros with the third-person camera when walking near walls.

**Comment by TranquilMarmot** [comment]
🤔 Maybe it's actually this issue? https://github.com/godotengine/godot/issues/69771

**Comment by ramokz** [comment]
I've seen this behavior since forever, including in some third-person demo projects.

Had a look at [GDQuest's recent Third-Person Controller demo project](https://www.gdquest.com/tutorial/godot/3d/third-person-character-controller/) and think there's a simple mitigation fix for it.
They're essentially adding a `SeparationRayShape3D` as the `Shape` with a Length of `0.5m` and `Slide on Slope` being enabled.

![image](https://github.com/user-attachments/assets/7d7cf57c-8287-461a-b9a9-0ae2b2f0b951)

There's still the odd frame here and there where it glitches through a collider, but it's significantly better by comparison, as it now also doesn't remain within the environment like before. At least, I have trouble replicating that behavior with the above change.

> 🤔 Maybe it's actually this issue? https://github.com/godotengine/godot/issues/69771

I did suspect that it was either an issue with the camera interpolating to the `PCam3D`, which might have been causing it, or a general problem. Given the examples the folks in the thread have shown mimics the issue here, I'm inclined to suspect there's something with how the `SpringArm3D` detects collision that causes the issue to occur — at least by default.
The MRP project from the issue did not get better by applying the `SeparationRayShape3D` approach above, so it's overall a bit odd…

**Comment by ramokz** [comment]
> It seems like a more "custom" solution might be necessary, at least until the bug in the engine is fixed. Not sure if that's something that would go into Phantom Camera since it's kind of a workaround.

Yeah, I'm not sure if that is something the addon should, let alone could, fix if the issue is present in typical `SpringArm3D` setups, or at least when minor changes are made to said setups. The addon just leverages that system, after all. So think coming up with a custom solution won't be worth the effort, as the engine should ideally make the necessary changes to make the `SpringArm3D` node.

**Comment by ramokz** [comment]
So the issue here might be more of a Godot related issue than an addon one?

**Comment by TranquilMarmot** [comment]
I think there are actually a couple of different issues at play here 😵 

I'm working through an in-engine fix for https://github.com/godotengine/godot/issues/69771 where the spring arm can sometimes go into a physics object.

I should have a PR open for it later today, but I imagine there will be a _lot_ of back-and-forth with the engine maintainers before it's merged and even then who knows if it will be in 4.4 or 4.5.

---

Another issue is the shape cast that `SpringArm3D` does.

If you look at the spring arm code in the engine you can see that it actually has some _drastically_ different behavior based on its children (I think this is kind of an anti-pattern, and honestly it was very surprising to me)

- If a "shape" is set, that's used as the shape to cast
- Otherwise, if a `Camera3D` is a _direct_ child of the `SpringArm3D`, a pyramid shape representing the camera's frustrum is used
- If no `Camera3D` is found, it falls back to a ray cast

https://github.com/godotengine/godot/blob/61accf060515416da07d913580419fd8c8490f7b/scene/3d/physics/spring_arm_3d.cpp#L143-L165

Also worth noting that the "margin" value of the spring arm is currently ignored in all scenarios except the ray cast fallback.

☝️ I have a hunch that since with phantom camera, there's no _direct_ `Camera3D` child of the spring arm, so if no shape is set in the third-person `PhantomCamera3D` then it falls back to the ray cast which is known to have some issues with tunneling/slipping through geometry.

In messing around with `SpringArm3D`, it _does_ feel like the best setup for it is to use the pyramid shape from the camera. Using a specific shape fails in a lot of scenarios I've tested, especially the separation ray shape.

`Camera3D` exposes a [`get_pyramid_shape_rid()` function](https://docs.godotengine.org/en/stable/classes/class_camera3d.html#class-camera3d-method-get-pyramid-shape-rid) . 

I think that Phantom Camera should set the spring arm's shape to this when no shape is provided explicitly, that way the spring arm behavior in phantom camera matches the "usual" spring arm behavior when a `Camera3D` is a direct child.

Took a bit of fiddling but you can generate the shape like this:
```gdscript
func _get_camera_shape() -> Shape3D:
	var pyramid_shape_data = PhysicsServer3D.shape_get_data(
		get_viewport().get_camera_3d().get_pyramid_shape_rid()
	)

	var shape = ConvexPolygonShape3D.new()
	shape.points = pyramid_shape_data

	return shape
```

☝️ I imagine that shape will need to be updated any time the camera's parameters are changed (i.e. near, far, FOV). As far as I could tell, there's no way to get a specific instance of a `Shape3D` from a given RID and the camera only has a way to get the RID of the shape.

Here's what it looks like with the fix I'm working on for https://github.com/godotengine/godot/issues/69771 _and_ using the camera's pyramid shape (I hacked `phantom_camera_3d.gd` to always use the shape from the above snippet):

https://github.com/user-attachments/assets/d1da26d5-f66d-4f88-ad70-b28d7bd5c496

It works okay with a sphere shape as well, as long as the sphere itself is small enough, but using the camera shape here seems to yield the most "expected" result.

---

A third issue I've run into is when the third-person phantom camera has a follow offset in the `x` direction.

When a `SpringArm3D` has an offset in the `x` direction, you run into an issue where it will go through walls.

https://github.com/user-attachments/assets/d2baa5ee-171e-49a7-b489-e3bfa780bde6

Since the spring arm itself is off to this side, this actually makes sense because the shape cast itself is actually happening _inside_ of the collision shape.

☝️ I think this issue is something that should **not** be fixed in Phantom Camera or in-engine and will require a custom solution for anybody that wants to use an offset spring arm. Something like: Check if you're close to a wall and change the offset accordingly.

**Comment by ramokz** [comment]
Fantastic findings!

Had no idea the `SpringArm3D` node internally changed the `Shape` to something very specific if a `Camera3D` was a child of it. Agree that it feels like a bit of an antipattern, as one would have assumed it wouldn't do anything behind-the-scenes, as that is never implied to the user.

> I think that Phantom Camera should set the spring arm's shape to this when no shape is provided explicitly, that way the spring arm behavior in phantom camera matches the "usual" spring arm behavior when a Camera3D is a direct child.

Agree, think it's something the addon should be doing as well,
Thank you for digging through the code and providing a sample of the setup. Will do some tests with it over the coming days. But by the sounds of it, it will greatly improve the out-of-the-box results. 

> Since the spring arm itself is off to this side, this actually makes sense because the shape cast itself is actually happening inside of the collision shape.

> I think this issue is something that should not be fixed in Phantom Camera or in-engine and will require a custom solution for anybody that wants to use an offset spring arm. Something like: Check if you're close to a wall and change the offset accordingly.

It makes sense, though I would have imagined that there would be something in the `SpringArm3D` itself that prevents it from going through physics colliders. Doesn't seem like an uncommon scenario to run into.

**Comment by ramokz** [comment]
Proposed addition resolved in #416
